cranelift/filetests/src/zkasm_codegen.rs

```diff
@@ -231,3 +235,85 @@ fn optimize_labels(code: &[&str], func_index: usize) -> Vec<String> {
     }
     lines
 }
+
+// TODO: fix same label names in different functions
+pub fn compile_clif_function(func: &Function) -> Vec<String> {
+    let flag_builder = settings::builder();
+    let isa_builder = zkasm::isa_builder("zkasm-unknown-unknown".parse().unwrap());
+    let isa = isa_builder
+        .finish(settings::Flags::new(flag_builder))
+        .unwrap();
+    let mut comp_ctx = cranelift_codegen::Context::for_function(func.clone());
+    let compiled_code = comp_ctx
+        .compile(isa.as_ref(), &mut Default::default())
+        .unwrap();
+    let mut code_buffer = compiled_code.code_buffer().to_vec();
+    fix_relocs(
+        &mut code_buffer,
+        &func.params,
+        compiled_code.buffer.relocs(),
+    );
+    let code = std::str::from_utf8(&code_buffer).unwrap();
+    let mut lines: Vec<String> = code.lines().map(|s| s.to_string()).collect();
+
+    // TODO: I believe it can be done more beautiful way
+    let mut funcname = func.name.to_string();
+    funcname.remove(0);
+    funcname.push(':');
+    let mut res = vec![funcname];
+    res.append(&mut lines);
+    res
+}
+
+// TODO: this function should be much rewrited,
+// now it works for one very basic case
+pub fn build_test_zkasm(functions: Vec<Vec<String>>, invocations: Vec<Vec<String>>) -> String {
+    // TODO: use generate_preambule to get preambule
+    let preambule = "start:\n  zkPC + 2 => RR\n  :JMP(main)\n  :JMP(finalizeExecution)";
+    let mut main = vec![
+        "main:".to_string(),
+        "  SP - 1 => SP".to_string(),
+        "  RR :MSTORE(SP)".to_string(),
+    ];
+    for invoce in invocations {
+        // TODO: remove this .clone()
+        main.append(&mut invoce.clone());
```

Comment:
You should be able to use [`Vec::extend`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.extend) for this with something like
```rust
main.extend(invocations.iter().cloned())
```

Reply:
`main.extend()` expects `Vec<String>` while `invocations` is `Vec<Vec<String>>`. And I'm not sure but I guess this `.cloned()` performs same amount of copying as `.clone()`, isn't it? 

But
```
for invocation in invocations {
    main.extend(invocation);
}
```
works.